### PR TITLE
Pass specific dates in Search Console deep links

### PR DESCRIPTION
**Pull request by johnPhillips**
## Summary

Addresses issue #2286

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/search-console/components/dashboard-details/DashboardDetailsWidgetKeywordsTable.js
Comment (tofumatt) [inline assets/js/modules/search-console/components/dashboard-details/DashboardDetailsWidgetKeywordsTable.js:34]: Now that #2563 is merged these imports can be cleaned up and are the likely source of merge conflicts in the PR—this can now be imported directly as `CORE_USER` instead of `STORE_NAME as CORE_USER`.

File: assets/js/modules/search-console/util/report-date-range-args.js
Comment (tofumatt) [inline assets/js/modules/search-console/util/report-date-range-args.js:4]: 2020 felt like it would never end, but it did 😉 

```suggestion
 * Site Kit by Google, Copyright 2021 Google LLC
```

File: assets/js/modules/search-console/util/report-date-range-args.js
Comment (tofumatt) [inline assets/js/modules/search-console/util/report-date-range-args.js:20]:
```suggestion
 * Generates a date range object with the appropriate structure to pass to report selectors.
```

Let's be specific about the kind of objects and/or reports we're talking about—this docstring will lack a bit of context without that.

File: assets/js/modules/search-console/util/report-date-range-args.js
Comment (tofumatt) [inline assets/js/modules/search-console/util/report-date-range-args.js:39]: Why return this if either of the expected arguments are false-y? I'd expect this either to error or to return undefined, but not to return the original arguments.

File: assets/js/modules/search-console/util/report-date-range-args.test.js
Comment (tofumatt) [inline assets/js/modules/search-console/util/report-date-range-args.test.js:4]:
```suggestion
 * Site Kit by Google, Copyright 2021 Google LLC
```

File: assets/js/modules/search-console/util/report-date-range-args.js
Comment (tofumatt) [inline assets/js/modules/search-console/util/report-date-range-args.js:29]: This can also return `undefined`, so this should be updated to be:

```suggestion
 * @return {Object|undefined} Date range object.
```

File: assets/js/modules/search-console/util/report-date-range-args.test.js
Comment (tofumatt) [inline assets/js/modules/search-console/util/report-date-range-args.test.js:28]: This is nowhere in the ACs/IBs and seems unexpected to me. Let's have it return `undefined` if the args aren't present. Or use `invariant` to warn against improper usage—either works for me.

Since it's a new function that isn't used anywhere else it might be nicer to use `invariant` when the args aren't a valid format. 🤔

**Review by tofumatt — CHANGES_REQUESTED** [review]
Looks good for the most part, but there are a few merge conflicts, likely related to the `STORE_NAME` change in #2563.

Also: let's not return the arguments as mentioned when they don't conform to what we expect—instead we should either error or return `undefined`.

File: assets/js/modules/search-console/util/report-date-range-args.test.js
Comment (johnPhillips) [reply assets/js/modules/search-console/util/report-date-range-args.test.js:28]: Bit of a misunderstanding I think: I picked this up from https://github.com/google/site-kit-wp/commit/58101f479d9cec82fe81b86a4234edc3161e3a6d , and understood / thought that the function had been written to spec. When I came to write the tests I included this behaviour. I did think it was unusual, but I'll just ask next time 🙂

File: assets/js/modules/search-console/util/report-date-range-args.js
Comment (johnPhillips) [reply assets/js/modules/search-console/util/report-date-range-args.js:39]: It was a handover mixup, see my comment below 🙈

**Comment by tofumatt** [comment]
@googlebot I consent.

**Review by tofumatt — COMMENTED** [review]
Just caught a few `STORE_NAME` references that needed updating, but otherwise the code looks good. I updated them to reduce the back-and-forth on the issue 🙂 

I saw E2E here failing but I'm not sure if that's related to the changes as I've seen them fail unexpected elsewhere on Actions too today. I'll wait until they pass to move to merge review, but it looks good to me.

**Comment by johnPhillips** [comment]
Thanks @tofumatt - I've renamed the imports as you suggest 👍

**Comment by johnPhillips** [comment]
E2E tests are passing locally...🤔

**Comment by tofumatt** [comment]
Odd, I see this test failure locally:

Screenshot 2021-01-21 at 14 53 59

**Review by tofumatt — COMMENTED** [review]
At this point I'm thinking merging this might be blocked by #2674, because E2E tests that are unrelated and should be passing are failing. I'll leave this in Code Review now but mark the issue as blocked by #2674.

**Comment by tofumatt** [comment]
@googlebot I consent.

**Comment by aaemnnosttv** [comment]
@johnPhillips the main problem here is that your linting is failing which is causing many other jobs to fail because it errors during the build.

**Comment by johnPhillips** [comment]
I rebased and fixed the linting issue (no idea how that crept in there, as I do check before I push). Looks like there is still an E2E test failing here. I'll try to get to this today and investigate further 👍

**Comment by aaemnnosttv** [comment]
> Looks like there is still an E2E test failing here

That doesn't look related to your changes here and seems to be one of the remaining instabilities. I've triggered it to run again after which it will _probably_ pass 😄  Usually legitimate E2E failures cause them all to fail, otherwise it's almost always a stability thing.

**Review by tofumatt — APPROVED** [review]

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

File: assets/js/modules/search-console/util/report-date-range-args.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/util/report-date-range-args.js:29]: As a function which expects the return value from the `getDateRangeDates` selector, we should require this to be an object since that selector will always return one (it never returns `undefined`).

File: assets/js/modules/search-console/components/dashboard-details/DashboardDetailsWidgetKeywordsTable.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/components/dashboard-details/DashboardDetailsWidgetKeywordsTable.js:32]: I realize @tofumatt raised this in his review https://github.com/google/site-kit-wp/pull/2634#discussion_r561120025, but our convention has always been to leave the `STORE_NAME` unaliased within the scope of the module's datastore and components. If this is confusing, we can revisit it in a dedicated issue to iterate on this, but we need to keep this for consistency.

TL;DR – modules or core components importing _their own_ `STORE_NAME` should do so using the generic `STORE_NAME` constant. We only used the named variant for referencing stores external to the current context.

This will reduce the size of the diff here as well.

File: assets/js/modules/search-console/components/dashboard-details/DashboardDetailsWidgetKeywordsTable.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/components/dashboard-details/DashboardDetailsWidgetKeywordsTable.js:34]: Just to reinforce the previous comment, this is correct.

File: assets/js/modules/search-console/util/report-date-range-args.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/util/report-date-range-args.js:40]: We should avoid using `undefined` in utility classes like this as a general rule regarding invalid input.

We primarily use `undefined` intentionally when we want to use a default argument (usually in another function), or to mean that some data is not resolved yet (i.e. `! hasFinishedResolution`).

In this case, I think raising an error is fine since it should never happen anyways and `getDateRangeDates` _always_ returns an object with these keys, so an error would indicate improper use.

```js
invariant( isValidDateString( startDate ), 'a valid startDate is required.' );
invariant( isValidDateString( endDate ), 'a valid endDate is required.' );
```

cc: @tofumatt

File: assets/js/modules/search-console/util/report-date-range-args.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/util/report-date-range-args.js:47]: Minor nitpick, but let's avoid creating unnecessary temporary variables before returning.
```suggestion
	return {
		start_date: startDate.replace( /-/g, '' ),
		end_date: endDate.replace( /-/g, '' ),
	};
```

File: assets/js/modules/search-console/util/report-date-range-args.test.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/util/report-date-range-args.test.js:46]: While this works fine, using a variable for all inputs somewhat obscures the readability of the test.
That is, variables are great for describing what it represents and using for default valid values throughout the tests, but for cases where we want to bring attention to the expectation, it helps to use literal values.
E.g.
```suggestion
			const { startDate, endDate } = allArgs;
			const result = generateDateRangeArgs( { startDate: '2020-12-31', endDate } );

			expect( result.start_date ).toBe( '20201231' );
```

File: assets/js/modules/search-console/util/report-date-range-args.test.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/util/report-date-range-args.test.js:25]: Since this variable is only destructured throughout, it would be cleaner to simply define `startDate` and `endDate` individually. It would also prevent these values from potentially being mutated and leaking between tests (not really a risk here but you get the idea).

This isn't a strict rule but a pattern you will likely see throughout the tests which has worked well for us.

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @johnPhillips – I see there was a lack of clarity around how the utility should work regarding invalid input, as well as confusion around the store name import.

I've left a few comments below to address, but it's all mostly small things. I left a few suggestions around the tests wells but nothing major.

All in all, this looks good just a handful of things to clean up then this should be good to go!

File: assets/js/modules/search-console/components/dashboard/GoogleSitekitSearchConsoleDashboardWidget.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/components/dashboard/GoogleSitekitSearchConsoleDashboardWidget.js:48]:
```suggestion
import { STORE_NAME, DATE_RANGE_OFFSET } from '../../datastore/constants';
```

File: assets/js/modules/search-console/components/dashboard/GoogleSitekitSearchConsoleDashboardWidget.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/components/dashboard/GoogleSitekitSearchConsoleDashboardWidget.js:64]:
```suggestion
	const propertyID = useSelect( ( select ) => select( STORE_NAME ).getPropertyID() );
	const isDomainProperty = useSelect( ( select ) => select( STORE_NAME ).isDomainProperty() );
```

File: assets/js/modules/search-console/components/dashboard/GoogleSitekitSearchConsoleDashboardWidget.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/components/dashboard/GoogleSitekitSearchConsoleDashboardWidget.js:76]:
```suggestion
	const searchConsoleDeepLink = useSelect( ( select ) => select( STORE_NAME ).getServiceURL(
```

File: assets/js/modules/search-console/util/report-date-range-args.test.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/util/report-date-range-args.test.js:27]: This works but Jest provides a [`toThrow`](https://jestjs.io/docs/en/expect#tothrowerror) assertion which makes this a bit cleaner:

```suggestion
			expect(
				() => generateDateRangeArgs( { startDate: '2020-12-18' } )
			).toThrow( 'A valid endDate is required' )
```

Note the value for matching the exception message is optional and supports various methods.
So, not that we can't/shouldn't use try/catch for this, but we should use `toMatch` where possible (this is where we use elsewhere as well).

File: assets/js/modules/search-console/util/report-date-range-args.test.js
Comment (aaemnnosttv) [inline assets/js/modules/search-console/util/report-date-range-args.test.js:33]:
```suggestion
			expect(
				() => generateDateRangeArgs( { endDate: '2021-01-14' } )
			).toThrow( 'A valid startDate is required' );
```

**Review by aaemnnosttv — COMMENTED** [review]
This one is super close, just a few minor details; they all have suggestions so I'll apply them now.

**Review by aaemnnosttv — DISMISSED** [review]
This worked great in my testing – thanks @johnPhillips !

File: assets/js/components/PostSearcher.js
Comment (aaemnnosttv) [inline assets/js/components/PostSearcher.js:45]: This is weird because it's already in `develop` but shows as a change... 🤔 
https://github.com/google/site-kit-wp/blob/01e0edf785ba0ab381c212359d8da885be1e6620/assets/js/components/PostSearcher.js#L45-L63

**Review by aaemnnosttv — APPROVED** [review]

**Pull request merged**
